core.js

```diff
@@ -60,16 +44,120 @@ function _check(buffer, headers, options) {
 }
 
 export async function fileTypeFromTokenizer(tokenizer) {
-	try {
-		return new FileTypeParser().parse(tokenizer);
-	} catch (error) {
-		if (!(error instanceof strtok3.EndOfStreamError)) {
-			throw error;
-		}
+	return new FileTypeParser().fromTokenizer(tokenizer);
+}
+
+export class TokenizerPositionError extends Error {
+	constructor(message = 'Tokenizer position changed without detecting a file type.') {
+		super(message);
+		this.name = 'TokenizerPositionError';
 	}
 }
 
-class FileTypeParser {
+export class FileTypeParser {
+	constructor(options) {
+		this.detectors = options?.customDetectors;
+
+		this.runCustomDetectors = this.#runCustomDetectors.bind(this);
+		this.fromTokenizer = this.fromTokenizer.bind(this);
+		this.fromBuffer = this.fromBuffer.bind(this);
+		this.parse = this.parse.bind(this);
+	}
+
+	async #runCustomDetectors(tokenizer) {
+		const {detectors} = this;
+		const initialPosition = tokenizer?.position;
+
+		for (const detector of detectors || []) {
+			const fileType = await detector(tokenizer);
+			if (fileType) {
+				return fileType;
+			}
+
+			if (initialPosition !== tokenizer?.position) {
+				throw new TokenizerPositionError();
```

Comment:
As discussed
```suggestion
				return undefined;
```

Reply:
In my understanding we agreed on not throwing the error to the user.
 The error is caught in the `FileTypeParser.fromTokenizer(...)`, which blends in nicely to the existing logic.
 
```
 		} catch (error) {
			if (!(error instanceof strtok3.EndOfStreamError) && !(error instanceof TokenizerPositionError)) {
				throw error;
			}
		}
```
What's the problem with this approach?